pkg/querier/ingester_querier.go

```diff
@@ -76,23 +89,36 @@ func (q *Querier) selectTreeFromIngesters(ctx context.Context, req *querierv1.Se
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
 
-	responses, err := forAllIngesters(ctx, q.ingesterQuerier, func(ctx context.Context, ic IngesterQueryClient) (clientpool.BidiClientMergeProfilesStacktraces, error) {
-		return ic.MergeProfilesStacktraces(ctx), nil
-	})
+	var responses []ResponseFromReplica[clientpool.BidiClientMergeProfilesStacktraces]
+	if plan != nil {
+		responses, err = forAllPlannedIngesters(ctx, q.ingesterQuerier, plan, func(ctx context.Context, ic IngesterQueryClient, hints *ingestv1.Hints) (clientpool.BidiClientMergeProfilesStacktraces, error) {
+			return ic.MergeProfilesStacktraces(ctx), nil
+		})
+	} else {
+		responses, err = forAllIngesters(ctx, q.ingesterQuerier, func(ctx context.Context, ic IngesterQueryClient) (clientpool.BidiClientMergeProfilesStacktraces, error) {
+			return ic.MergeProfilesStacktraces(ctx), nil
+		})
+	}
 	if err != nil {
 		return nil, connect.NewError(connect.CodeInternal, err)
 	}
 	// send the first initial request to all ingesters.
 	g, gCtx := errgroup.WithContext(ctx)
-	for _, r := range responses {
-		r := r
+	for idx := range responses {
+		r := responses[idx]
+		hints, ok := plan[r.addr]
+		if !ok && plan != nil {
+			return nil, connect.NewError(connect.CodeInternal, fmt.Errorf("no hints found for replica %s", r.addr))
+		}
+
```

Comment:
Regarding the nil check, not too sure I can follow, why I need to do the plan-nil check before?

When a plan is not nil, we will only query ingesters which are part of the plan (`forAllPlannedIngesters`), so we should never be in the situation to hit a replica not part of the planreally, so I think it should be an error.

Reply:
You're right, I overlooked that. Indeed, now I see that this may only happen if `forAllPlannedIngesters` removed the replica from plan, or got a response from unexpected one, which would be a bug

RE nil check: never mind, it should be okay